### ISSUE TITLE
lake clean can delete files outside of its build directories

### ISSUE THREAD

**Issue by Julian** [leanprover/lean4#2731]
### Prerequisites

* [x] Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
* [ ] Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

`lake clean` seems to sometimes delete files other than its build output, including entire projects.

After running it a few times, it successively deleted a number of dependency directories (of Mathlib4 in my case) but then deleted the entire Mathlib4 repository I was working in.

### Context

Two instances of this were reported on Zulip, a first one [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/lake.20exe.20cache.20get.20errors) and a second one I encountered myself [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/lake.20build'ing.20mathlib.20HEAD)

### Steps to Reproduce

It's not fully clear what causes lake to get in this situation. The first Zulip thread seems to suggest the issue is related to a faulty `lakefile.olean` file from an incorrect version of Lean.

**Expected behavior:** All build outputs are deleted, including `lake-packages`, `build`, and `lakefile.olean`. No other files are touched, and I'd expect `lake` to never attempt to write or modify any file other than its build output directories.
**Actual behavior:** The whole repository was deleted.

### Versions

[Output of `lean --version` in the folder that the issue occured in]

```
⊙  lean --version                                                                                                                                                                                                                                                                                                   julian@Airm
Lean (version 4.2.0, Release)
```

This is rc4 to be clear, which I see doesn't appear in the output.

The other Zulip thread appears to have been on 4.2.0 rc3.

[OS version] macOS Sonoma 14.0 

### Additional Information

#2592 seems tangentially related.

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by tydeu** [comment]
The problem here is likely that since `rc4` reverted the ABI incompatibility with `rc1` present in `rc3`, `rc4` is now ABI incompatible with `rc3` (in the same way).

**Comment by tydeu** [comment]
With the proper disclaimer now on the [`rc4`](https://github.com/leanprover/lean4/releases/tag/v4.2.0-rc4) release of how to safely update to it to avoid this issue, I am going to close this. Future ABI incompatibility issues should be solved by #2749. Going forward, features like #2741 and #2740 should hopefully be added in future to further avoid this popping up again.
